matrix_sdk_base/src/models/room.rs

```diff
@@ -631,28 +632,79 @@ impl Room {
     #[cfg(feature = "messages")]
     #[cfg_attr(docsrs, doc(cfg(feature = "messages")))]
     pub fn handle_message(&mut self, event: &AnyMessageEventStub) -> bool {
-        let message = MessageWrapper::clone_into_any_content(event);
-        self.messages.push(message)
+        self.messages.push(FullOrRedactedEvent::Full(event.clone()))
     }
 
     /// Handle a room.redaction event and update the `MessageQueue`.
     ///
-    /// Returns true if `MessageQueue` was updated. The effected message event
-    /// has it's contents replaced with the `RedactionEventContents` and the whole
-    /// redaction event is added to the `Unsigned` `redacted_because` field.
+    /// Returns true if `MessageQueue` was updated. The affected message event
+    /// has it's contents replaced with the `Redacted[event_name]EventContent` and the whole
+    /// redaction event is added to the `unsigned` `redacted_because` field.
     #[cfg(feature = "messages")]
     #[cfg_attr(docsrs, doc(cfg(feature = "messages")))]
-    pub fn handle_redaction(&mut self, event: &RedactionEventStub) -> bool {
+    pub fn handle_redaction(&mut self, redacted_event: &RedactionEventStub) -> bool {
         if let Some(msg) = self
             .messages
             .iter_mut()
-            .find(|msg| event.redacts == msg.event_id)
+            .find(|msg| &redacted_event.redacts == msg.event_id())
         {
-            msg.content = AnyMessageEventContent::RoomRedaction(event.content.clone());
-
-            let redaction =
-                redaction_event_from_redaction_stub(event.clone(), self.room_id.clone());
-            msg.unsigned.redacted_because = Some(EventJson::from(redaction));
+            match msg.deref_mut() {
+                FullOrRedactedEvent::Full(event) => match event {
+                    AnyMessageEventStub::RoomMessage(event) => {
+                        let MessageEventStub {
+                            content,
+                            event_id,
+                            sender,
+                            origin_server_ts,
+                            mut unsigned,
+                        } = event.clone();
+                        // TODO can redacted events that are stripped or have no room_id be
+                        // added to the `redacted_because` field?
```

Comment:
Stripped state events are received in the invited rooms, if so I don't think that we're getting new events there, nor redactions.

A for no `room_id`, events that we get using the sync lack the room_id, so this is obviously yes, but we can convert if needed. So all events that end up here should lack the room id or am I missing something?

Reply:
The way it is now all the events that end up in `redacted_because` have a `room_id` field. The only purpose of `redaction_event_from_redaction_stub` is to add the `room_id` The problem is possibly more in ruma. Ruma requires the `redacted_becasuse` field to be an event with a `room_id` I was wondering if you had any idea if `redacted_because` should be able to accept events with or without the `room_id` field?

Reply:
@poljar Unfortunately, whether "room id stripping" is propagated to `redacted_because` is not documented anywhere (or am I missing sth.?) and nobody in #matrix-spec answered my question about it either. Thinking about it though, redactions probably can't cross room boundaries, can they? If they can't, there's no point in there being a room_id field inside `redacted_because` at all.

> Stripped state events are received in the invited rooms

Yes

> I don't think that we're getting new events there, nor redactions.

Why is that?

Reply:
You can get redacted events through the sync or through any other endpoint that returns events so I think it will depend on that.

If I take a look at what Synapse sends out in a sync response I get this:

```
"redacted_because": {
    "content": {},
    "event_id": "$YxS81Y5hYOwi3L4QFdw9sNwxOXmeoT7a_zxSVVsvpi0",
    "origin_server_ts": 1594635906680,
    "redacts": "$8B8WKPysjCHv56COFAfyZ8kHSZxmO-HTf5_gACeVr20",
    "sender": "@example:localhost",
    "type": "m.room.redaction",
    "unsigned": {
        "age": 14133974
    }
}
```

If I take a look at what we get when we use the `/rooms/${room_id}/messages` endpoint:

```
"redacted_because": {
    "age": 7180657888,
    "content": {},
    "event_id": "$iwtdH9vFhya4YPzPrtooBMVHZcW-1cbLErWMAFDkYhE",
    "origin_server_ts": 1587469588786,
    "redacts": "$Sb8IEHOWv1mwUdDVP0F1Vd7BVdLJ6wbu6PgmAeALijQ",
    "room_id": "!ocStHylJqlUbTjILPE:localhost",
    "sender": "@example:localhost",
    "type": "m.room.redaction",
    "unsigned": {
         "age": 7180657888
    },
    "user_id": "@example:localhost"
}
```

So it seems like the same rules apply to the room_id field as usual.

Reply:
Okay. If you could open an issue about that on Ruma that would be great, otherwise I can write one up later.

Reply:
> @poljar Unfortunately, whether "room id stripping" is propagated to `redacted_because` is not documented anywhere (or am I missing sth.?) and nobody in #matrix-spec answered my question about it either. Thinking about it though, redactions probably can't cross room boundaries, can they? If they can't, there's no point in there being a room_id field inside `redacted_because` at all.

They can't cross room boundaries.


> > I don't think that we're getting new events there, nor redactions.
> 
> Why is that?

Indeed I should check that, my thinking was that the invited room gives you a checkpoint in the room state and you should not receive new updates since that might leak metadata which you should only have access to once you join. It's quite unclear in the spec on how this should behave.

Reply:
Even with it just giving you a checkpoint, that can include redacted events, no? E.g. a new room with the creation event redacted.

Reply:
Indeed, I'll check this out.

Reply:
https://github.com/ruma/ruma/issues/126

Reply:
So I checked this out and redactions don't seem to propagate to the `invited_state` at all, even if I redact some of the initial room creation events, nor do they seem to end up there if you dis-invite the member and invite again. 

The redacted event just ends up missing in the `invited_state` and we don't seem to get updates when we do redact some of them that end up in the `invited_state`.

Reply:
@poljar As long as this isn't documented though, I wouldn't want to assume all of that. It would probably even be more work to have redacted messages only in `AnyMessageEvent`, `AnySyncMessageEvent`, `AnyStateEvent` and `AnySyncStateEvent`, but leave them out of `AnyStrippedStateEvent` (but of course if that's what the spec says, we will implement it).

Reply:
Sure, if it's more work to not have them there let's allow them. The spec is quite ambiguous about that and I assumed that it'll be more work to have them there.